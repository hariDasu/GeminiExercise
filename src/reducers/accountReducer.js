import * as types from "../types";

export const initialAccountState = {
  transactionHistory: {},
  fromAddress: "",
  balance: 0,
  success: null,
  error: null,
  isLoading: null,
};

export const account = (state = initialAccountState, action) => {
  switch (action.type) {
    case types.accountTypes.ACCOUNT_HISTORY_REQUEST:
      return {
        ...state,
        fromAddress: action.address,
        success: false,
        error: false,
        isLoading: true,
      };
    case types.accountTypes.ACCOUNT_HISTORY_SUCCESS:
      return {
        ...state,
        isLoading: false,
        success: true,
        error: false,
        transactionHistory: action.transactions,
        balance: action.balance,
      };

    case types.accountTypes.ACCOUNT_HISTORY_FAILURE:
      return {
        ...state,
        isLoading: false,
        success: false,
      };

    default:
      return state;
  }
};
