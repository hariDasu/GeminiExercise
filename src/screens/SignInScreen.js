import React, { useState, useCallback, useEffect } from "react";
import { StyleSheet, View, TextInput } from "react-native";
import { GenericButton } from "../components";
import { useDispatch, useSelector } from "react-redux";
import * as actions from "../actions";

export const SignInScreen = ({ navigation }) => {
  const dispatch = useDispatch();
  const transactionHistory = useSelector(
    (state) => state.account.transactionHistory
  );
  const [address, setAddress] = useState("");
  const signInDisabled = address === "";

  const fetchTransactionHistory = useCallback(() => {
    dispatch(actions.getTransactionsRequest(address));
    return navigation.navigate("TransactionHistory", { transactionHistory });
  }, [dispatch, address, navigation, transactionHistory]);

  return (
    <View style={styles.containerStyle}>
      <TextInput
        style={styles.input}
        onChangeText={setAddress}
        value={address}
        placeholder="Enter address"
        keyboardType="default"
        testID={"addressField"}
      />
      <GenericButton
        onPress={fetchTransactionHistory}
        disabled={signInDisabled}
        title="Sign in..."
        testID={"signInButton"}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  containerStyle: {
    height: 300,
    alignItems: "center",
    justifyContent: "space-evenly",
  },
  input: {
    height: 40,
    width: "60%",
    borderTopWidth: 0,
    borderBottomWidth: 1,
  },
});
