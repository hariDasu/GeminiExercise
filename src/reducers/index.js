import { combineReducers } from "redux";
import { account } from "./accountReducer";

export default combineReducers({
  account,
});
