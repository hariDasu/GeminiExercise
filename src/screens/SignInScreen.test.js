import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { NavigationContainer } from "@react-navigation/native";

import { SignInScreen } from "./SignInScreen";
import rootReducer from "../reducers";
import { initialAccountState } from "../reducers/accountReducer";
import { act } from "react-test-renderer";

jest.mock("react-native-chart-kit", () => {});
jest.mock("react-native-gesture-handler", () => {});

function SignInComponent() {
  const store = configureStore({
    reducer: rootReducer,
    preloadedState: { account: initialAccountState },
  });
  return (
    <NavigationContainer>
      <Provider store={store}>
        <SignInScreen />;
      </Provider>
    </NavigationContainer>
  );
}
describe("SignIn", () => {
  test("Renders correctly", () => {
    const { toJSON } = render(<SignInComponent />);
    expect(toJSON()).toMatchSnapshot();
  });
  test("Navigates to transaction history", async () => {
    const navigate = jest.fn();
    const { getByTestId } = render(
      <SignInComponent navigation={{ navigate }} />
    );
    const addressInput = getByTestId("addressField");
    await act(async () => {
      await waitFor(() => {
        fireEvent.changeText(getByTestId("addressField"), "Henry");
      });
      expect(addressInput.props.value).toBe("Henry");
    });
    //TODO: properly initialize react navigation mocking with redux
    // fireEvent(getByTestId("signInButton"), "press");
    // expect(navigate).toHaveBeenCalledWith("TransactionHistory");
  });
});
