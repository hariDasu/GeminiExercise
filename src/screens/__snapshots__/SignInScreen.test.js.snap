// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Renders correctly 1`] = `
Array [
  <View
    style={
      Object {
        "alignItems": "center",
        "height": 300,
        "justifyContent": "space-evenly",
      }
    }
  >
    <TextInput
      keyboardType="default"
      onChangeText={[Function]}
      placeholder="Enter address"
      style={
        Object {
          "borderBottomWidth": 1,
          "borderTopWidth": 0,
          "height": 40,
          "width": "60%",
        }
      }
      testID="addressField"
      value=""
    />
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#999999",
          "borderRadius": 25,
          "borderWidth": 0.5,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "width": "80%",
        }
      }
      testID="signInButton"
    >
      <Text
        style={
          Array [
            Object {
              "color": "#212121",
              "fontSize": 16,
              "letterSpacing": 1.39,
              "lineHeight": 20,
            },
            Object {
              "color": "#212121",
            },
          ]
        }
      >
        Sign in...
      </Text>
    </View>
  </View>,
  ";",
]
`;
